File: tareas/2/24190112/14_Desviacion_estandar/Desviacion_estandar.py
import re
import math

def Leer(archivo_entrada):
    with open(archivo_entrada, encoding="utf-8") as entrada:
        texto = entrada.read()
    return texto

def Extraer(texto):
    numeros = re.findall(r'-?\d+', texto)
    numeros = [float(num) for num in numeros]
    return numeros

def Escribir(archivo_salida, datos):
    with open(archivo_salida, "w", encoding="utf-8") as salida:
        resultado = Calcular_desviacion_estandar(datos)
        salida.write(f"La desviación estándar de la lista de números dada es: {resultado:.2f}")
        
#Para sumar en una lista que seria como el equivalente a un vector es mas fácil.

def Calcular_desviacion_estandar(datos):
    
    media = sum(datos) / len(datos)

    suma_cuadrados_diferencias = sum((dato - media)**2 for dato in datos)

    varianza = suma_cuadrados_diferencias / len(datos)

    desviacion_estandar = math.sqrt(varianza)

    return desviacion_estandar

def main():
    archivo_entrada = "entrada_ejercicio14.txt"
    archivo_salida = "salida_ejercicio14.txt"
    
    texto = Leer(archivo_entrada)
    datos = Extraer(texto)

    Escribir(archivo_salida, datos)

if __name__ == "__main__":
    main()

File: examenes/finales/2024-1_grupo-2/desviacion_estandar.cpp
#include <cmath>
#include <iostream>
#include <vector>
using namespace std;

long double Media(const vector<long double>& datos) {
    long double suma = 0.0L;
    for (long double dato : datos) suma += dato;
    return suma / datos.size();
}

long double DesviacionEstandar(const vector<long double>& datos) {
    long double media = Media(datos);
    long double suma = 0.0L;
    for (long double dato : datos) {
        long double delta = media - dato;
        suma += delta * delta;
    }
    return sqrt(suma) / datos.size();
}

int main(void) {
    vector<long double> v = {14.0L, 15.0L, 19.0L, 20.0L, 20.0L};
    cout << DesviacionEstandar(v) << endl;
    return 0;
}